File: 2024/08/23/39. Combination Sum/solution.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> temp;
        dfs(candidates,target,0,temp,0,res);
        return res;
    }
    void dfs(vector<int>& candidates, int target, int idx, vector<int>& temp, int total, vector<vector<int>>& res){
        if (total == target) {
            res.push_back(temp);
            return;
        }
        if (total > target || idx >= candidates.size()) {
            return;
        }
        temp.push_back(candidates[idx]);
        dfs(candidates,target,idx,temp,total+candidates[idx],res);
        temp.pop_back();
        dfs(candidates,target,idx+1,temp,total,res);
    }
};

File: 2024/08/23/39. Combination Sum/solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        candidates.sort()
        def dfs(first,target,path):
            if target == 0:
                result.append(path)
                return
            if target < 0:
                return
            for i in range(first,len(candidates)):
                dfs(i,target-candidates[i],path+[candidates[i]])
        dfs(0,target,[])
        return result
        